playback/src/audio_backend/gstreamer.rs

```diff
@@ -47,15 +54,15 @@ impl Open for GstreamerSink {
         let bufferpool = gst::BufferPool::new();
         let appsrc_caps = appsrc.get_caps().expect("couldn't get appsrc caps");
         let mut conf = bufferpool.get_config();
-        conf.set_params(Some(&appsrc_caps), 8192, 0, 0);
+        conf.set_params(Some(&appsrc_caps), 2048 * format.size() as u32, 0, 0);
         bufferpool
             .set_config(conf)
             .expect("couldn't configure the buffer pool");
         bufferpool
             .set_active(true)
             .expect("couldn't activate buffer pool");
 
-        let (tx, rx) = sync_channel::<Vec<u8>>(128);
+        let (tx, rx) = sync_channel::<Vec<u8>>(64 * format.size());
```

Comment:
This hasn't scaled the same way as the 8192 number above, is that intentional? Sorry for a maybe dumb Rust question but what actually is the value of `format.size()` here, for example in the case of `AudioFormat::S16`?

I'm not sure I understand this backend enough to see why this (seemingly random) value needs to scale but thinking out loud... With larger sample formats, `write_bytes()` is now called with more data each time, that pushes more `data` each time into the appsrc, its internal queue of 4096 fills up sooner, which means `appsrc.push_buffer()` blocks sooner, which I assume then causes this sync_channel to fill its buffer sooner and ultimately block the `send()`. So either the sync_channel size needs to scale with the sample format (as you have), or the appsrc's max-bytes needs to scale. I think that makes sense.

Reply:
Hi, so I originally got a fair bit of this code from Paul Lietar's (plietar) "gst" branch of librespot, which never got finished until I picked up the mantle, for several reasons:

1. It had a lot of major latency/synchronization issues that were only improved significantly in librespot by the introduction of the gapless playback feature.

2. The refactoring of the audio_sink.rs to multiple modular audio backends required significant porting work which I was the first person to undertake.

3. GStreamer bindings for Rust had significantly diverged from the version Paul had used in the abandoned gst branch. Some were more than just API synonyms and required re-thinking the code's architecture a little bit.

That said, a lot of the magic numbers in the code are holdovers from plietar's code, which is in and around this specific line: https://github.com/plietar/librespot/blob/gst/src/audio_sink.rs#L160

I recall specifically setting the sync_channel's size to 128 because I went from plietar's original `sync_channel::<Vec<u16>>(64)` to `sync_channel::<Vec<u8>>(128)`. You get the same number of bytes either way; it's 1 KiB, which is a convenient number for passing data between threads with both low latency and low overhead, without causing horrible thrashing.

The `sync_channel` literally ferries the PCM audio data, so increasing the size of it would basically add latency to the GStreamer playback backend for librespot. Decreasing the size of it would cause it to become full so often that you'd be spending most of your CPU time just dealing with the internals of the `sync_channel` as the buffer constantly fills and empties.

I guess it's an optimization problem, and plietar settled on a value that I haven't had the need to dispute. Basing it on the scaling of the `format.size()` makes sense to me; u16 is a 2-byte unsigned integer, so `format.size() == 2`, since it's in bytes. 

So the way he has it coded in this PR, if the input PCM data is 16-bit, we still get a `Vec<u8>` with a total size of 1 KiB; if the input data is 32-bit, the `Vec` increases in size to 2 KiB, which preserves the latency of the original code given a constant sample rate, because you literally need more data to handle the same number of samples, if each sample is larger.

`format.size()` is just the number of bytes per sample in the PCM datastream. A sample is a single value of an integral number type for a single channel of audio. Since one of the main goals of this PR is to increase the "width" of the samples we play with to increase processing headroom for volume and normalization, this makes a ton of sense to me based on my intermediate knowledge of audio DSP.

Reply:
If it's a question of low latency/overhead vs a little higher latency/overhead but better stability (as in less/no buffer underruns). I'd pick stability. Latency is not an issue since the audio doesn't really need to be sync'd to anything except maybe the time slider in the controlling app. As long as it's less than 1 sec you're good.

Overhead is kinda over my head,lol!!! If by overhead you're talking about memory and CPU usage (?) if your system doesn't have enough memory to store 1 sec of audio you're got other problems. (stereo 16-bit, 44.1 KHz PCM = 176 kbyte/sec) And a larger buffer actually uses less CPU up to a point. This again goes back to picking a potato and seeing what works on that lowest spec target device. A crazy small buffer may work great on a x86 system but could absolutely murder a little ARM board.

Reply:
>That said, a lot of the magic numbers in the code are holdovers from plietar's code,

Thanks for the reply, that's still helpful to know. I am sorry @roderickvd if this is extra noise in the grand scheme of merging your PR but since the code is changing here I think it makes sense to keep this discussion here also.

> You get the same number of bytes either way; it's 1 KiB, which is a convenient number for passing data between threads with both low latency and low overhead, without causing horrible thrashing.

I thought `<Vec<u8>>(128)` is 128 buffer entries with each entry being a `Vec<u8>` i.e. each entry varies in size depending on how many samples the vorbis decoder gave us in that particular VorbisPacket and how many bytes we're using to store each sample. The number of samples we get in each VorbisPacket seems to vary during playback (as low as 256, mostly 2048, and sometimes as high as 2944 samples in my 10-second random survey). So, when the appsrc's buffer (4096 bytes) goes full (after most writes!) and it blocks, sync_channel then starts filling its 128-entry buffer and can handle 128 more VorbisPackets (of whatever size) being written until it also must block and therefore blocks `Player` - which I think is part of the playback throttling dance sinks do to avoid #521. We could shrink that sync_channel right down and I think all that would happen is it'd block `Player` a little sooner.

The point being that the sync_channel's buffer depth isn't important here since it's not directly related to the sample size. We should scale appsrc `max-bytes` instead. In general I think it's usually better to do all your buffering in your GST pipeline because it will properly handle when they need to be flushed (although I don't think we take advantage of that here).

My question about `format.size()` was because I couldn't find the `size()` implemention for `AudioFormat` so I thought maybe it was some Rust magic I hadn't seen before... But it turns out I just hadn't pulled the latest branch changes and [I see it now](https://github.com/roderickvd/librespot/blob/hi-res-volume-control/playback/src/config.rs#L60)!

Which goes back to my previous comment but maybe I wasn't clear. The latest changes to scale these numbers don't match the old values. Specifically for S16 sample format, `format.size()` is 2 which means the GST buffers are now half the size we've currently got, they are too small and cause `'assertion failed: self.get_maxsize() >= size'` for S16 format (presumably also for the other formats).

Reply:
Thanks for the clarification; I totally misread the code there! Shows you how much I know about something I purportedly worked on, heh.

Out of curiosity, how does one reproduce that assertion failure? Is it happening always, only on certain songs, or only with certain gst playback pipelines?

Reply:
Totally understandable, It was a while ago. I can't remember how code works that I wrote last week! 

I tried a few random songs in the random playlist I had loaded up on my phone client. With the buffer size reduced to 4096 it happens virtually straight away, within the first 30 VorbisPackets. This is with the normal pipeline in the librespot gstreamer backend e.g. running `./target/debug/librespot -n Librespot --backend gstreamer`. I must admit, I don't know why the VorbisPacket size varies and why it takes the few specific values it does:
* 256 (512 bytes)
* 1152 (2304 bytes)
* 2048 (4096 bytes)
* 2944 (5888 bytes <- assert!)
* maybe others?

Reply:
> Thanks for the reply, that's still helpful to know. I am sorry @roderickvd if this is extra noise in the grand scheme of merging your PR but since the code is changing here I think it makes sense to keep this discussion here also.

@kingosticks on the contrary. This PR touches a lot of code, so the more reviews, the better.

> Which goes back to my previous comment but maybe I wasn't clear. The latest changes to scale these numbers don't match the old values. Specifically for S16 sample format, format.size() is 2 which means the GST buffers are now half the size we've currently got, they are too small and cause 'assertion failed: self.get_maxsize() >= size' for S16 format (presumably also for the other formats).

No, this is not the case. In master `dev` it's:

https://github.com/librespot-org/librespot/blob/56f1fb6dae081daa88b4429c3970174ea45a0c91/playback/src/audio_backend/gstreamer.rs#L50

and:

https://github.com/librespot-org/librespot/blob/56f1fb6dae081daa88b4429c3970174ea45a0c91/playback/src/audio_backend/gstreamer.rs#L110-L112

where `packet.samples(&self) -> &[i16]` and `as_bytes(&self) -> &[u8]`.

So for S16 this is unchanged and does not panic. And for S32 and F32 the capacity is doubled, likewise without panicking. At least not on my RPi 3B+. Have you compiled from the most recent https://github.com/librespot-org/librespot/pull/660/commits/770ea15498a0f1cfc7b9986f0954f4150258c29f?

> I tried a few random songs in the random playlist I had loaded up on my phone client. With the buffer size reduced to 4096 it happens virtually straight away, within the first 30 VorbisPackets. 

You're running `librespot` on your phone or do you mean you're using Spotify Connect from your phone UI to some other device? I can revert the buffer pool size to `4096 * format.size()` if that fixes this?

Reply:
Sorry, I guess it still wasn't clear because I was talking about both scaling numbers in my last comment.

Your latest commit (https://github.com/roderickvd/librespot/commit/770ea15498a0f1cfc7b9986f0954f4150258c29f) breaks S16 playback because it changes 
```
conf.set_params(Some(&appsrc_caps), 8192, 0, 0);
```
to
```
conf.set_params(Some(&appsrc_caps), 2048 * format.size() as u32, 0, 0);
```
Which, for S16, is 4096 bytes. This is half of 8192 and this is too small. Changing it to `4096 * format.size()` fixes it.

Reply:
>  For decoding video to fakesink, buffers must be aligned to 128 bytes and padded with 4096 bytes. The latter seems a requirement depending on the sink. Many other bells and whistles on GStreamer default to 4096. I'm taking this as anecdotal evidence for audio and other sinks.

In that (fictional) example, it's the fictional video decoder that has those requirements. They are just examples. There's nothing special about the number 4096 in bufferpools. 

> First, <Vec<u8>>(64 * format.size()) is the minimum size channel. Go any lower and it will panic.

What panic do you get? I reduced it to 16 earlier without any problems. I'm not suggesting we change it but I don't understand why it has any dependency on the format size. 

I still think we should be scaling `max-bytes` to scale this all properly and get the same blocking behaviour regardless of sample size. From https://gstreamer.freedesktop.org/documentation/app/appsrc.html?gi-language=c
> “max-bytes” guint64
> The maximum amount of bytes that can be queued internally. After the maximum amount of bytes are queued, appsrc will emit the "enough-data" signal.
>
> “block” gboolean
> When max-bytes are queued and after the enough-data signal has been emitted, block any further push-buffer calls until the amount of queued bytes drops below the max-bytes limit.

Reply:
> > First, <Vec>(64 * format.size()) is the minimum size channel. Go any lower and it will panic.
> 
> What panic do you get? I reduced it to 16 earlier without any problems. I'm not suggesting we change it but I don't understand why it has any dependency on the format size.

I can reproduce this every time: https://github.com/librespot-org/librespot/pull/660#issuecomment-800063904.

> I still think we should be scaling `max-bytes` to scale this all properly and get the same blocking behaviour regardless of sample size. From https://gstreamer.freedesktop.org/documentation/app/appsrc.html?gi-language=c

You're the expert here, unless others say otherwise I can scale both the bufferpool size and `max-bytes`. So what should they be, `4096 * format.size()` for the pool and `2048 * format.size()` for `max-bytes`? Let me know and I'll push a commit.

Reply:
Perhaps that `set_size()` panic is somehow hardware dependent. I can't reproduce and it's concerning since I also don't understand why it would happen. I wish I understood why the decoder generated different amounts of data because maybe that's related. Maybe you can get in a situation where it gives you more than 8KB and playing with the channel buffer on different hardware makes that come out. What hardware are you using?

I think those two suggestions are good.

Reply:
> I wish I understood why the decoder generated different amounts of data because maybe that's related.


Is it happening at the beginning of a song? I ask because Gstreamer does not handle buffering on it's own. What I mean is that it's up to the application to pause the player if/when buffering happens.

Reply:
https://gstreamer.freedesktop.org/documentation/application-development/advanced/buffering.html?gi-language=c

Reply:
> What hardware are you using?

Myself a Raspberry Pi 3 Model B+ in 32 bit. As for @JasonLG1979 I'm not sure if it was his Pi Zero or desktop?

Reply:
I tested Gstreamer on my desktop.

Reply:
OK, I was able to repro the issue building this PR branch with MinGW64 on Windows 10. 

I just doubled the bufferpool size on gstreamer.rs:L57 from

`conf.set_params(Some(&appsrc_caps), 2048 * format.size() as u32, 0, 0);`

to 

`conf.set_params(Some(&appsrc_caps), 4096 * format.size() as u32, 0, 0);`

This is the only change of a "magic number" anywhere in the file that fixes the assertion (the one that kingosticks reported). Messing with the `tx` / `rx`, the appsrc's `max-bytes`, etc. has no effect.

My output pipeline is `! audioconvert ! pitch pitch=0.95 ! audioconvert ! audioresample ! autoaudiosink` if that makes any difference, but it really shouldn't (the `audioconvert`s should handle any necessary sample format conversion from the source data regardless of what it is, and `pitch` can handle all sample rates; once it's processed the `audioconvert` coupled to an `audioresample` will guarantee that the `autoaudiosink` can negotiate caps that the hardware is happy with in 99.9999999% of cases, unless the output sink has a bug.)

Reply:
@allquixotic There really should be a queue2 element in there some where to handle buffering instead of creating a static buffer and hoping for the best.

Reply:
> 
> 
> @allquixotic There really should be a queue2 element in there some where to handle buffering instead of creating a static buffer and hoping for the best.

The bufferpool is the means by which we copy data from the librespot thread that gives us data in the `SinkAsBytes` (see gstreamer.rs:L122) into GStreamer without blocking to wait for data. We can't really eliminate all threading code in gstreamer.rs, either, because GStreamer has to run a main event loop to work at all. It can integrate with many other programs' main event loops and share a thread on the event loop for certain compatible event loop implementations, but I don't know how or if we could integrate it into librespot's main loop, assuming our goal was to eliminate all the threading code in gstreamer.rs.

I added queues right after the appsrc and right before my autoaudiosink, and with the original source code as-is in roderickd's branch, I still get the assertions. I believe as long as we are using this method to copy data into gstreamer, we have to ensure the GstBufferPool is at least as large as the maximum amount of data that can be transmitted by librespot to gstreamer in a single batch.

Reply:
try a pipeline of appsrc > queue2 > audioconvert > audioresample > autoaudiosink. Set use-buffering to true in queue2 and actually handle buffering. I honestly think that's the problem. As I've said before Gstreamer does not automatically handle buffering. It's more than happy to starve the buffer and crash. Also I don't know what it looks like before the PCM is handed to the backends but it will definitely crash if you try to start playback before the buffer is full.

Reply:
You can also use the `max-size-time` prop in queue2 and set it to 500000000 (0.5 secs) and not have to worry about calculating the size of buffer in bytes.

Reply:
> The bufferpool is the means by which we copy data from the librespot thread that gives us data in the `SinkAsBytes` (see gstreamer.rs:L122) into GStreamer without blocking to wait for data.

Well it's obviously very brittle. Are you making sure there's enough data to actually fill the buffer pool before you try to fill it?

Reply:
> 
> 
> try a pipeline of appsrc > queue2 > audioconvert > audioresample > autoaudiosink. Set use-buffering to true in queue2 and actually handle buffering. I honestly think that's the problem. As I've said before Gstreamer does not automatically handle buffering. It's more than happy to starve the buffer and crash. Also I don't know what it looks like before the PCM is handed to the backends but it will definitely crash if you try to start playback before the buffer is full.

To "actually handle buffering" I would have to listen for `GST_MESSAGE_BUFFERING` on the message bus in the GST mainloop, right? I know how to set up a bus message handler because this codebase already has a handler for EOS, but I don't have any good examples (in C or Rust) at hand of how to handle `GST_MESSAGE_BUFFERING` and use the queue2's buffers for buffering the data instead of a static `GstBufferPool`.

 The audio playback API in librespot is still designed so that it passes the PCM data to you as a "push" of a byte array (push as opposed to pull API), so the only way I'd be able to go from push to pull is if we rewrote the audio playback API or had some kind of shim in gstreamer.rs. 

Basically, what I'm saying is I don't understand how to "actually handle buffering" in an elegant way while using a dynamically sized buffer controlled by the queue instead of a statically allocated buffer. I can easily set the properties on a queue2 when building the pipeline, but I'm thinking the minimum changes to convert to using a queue2 here would be:

1. Put a queue2 in the pipeline behind the appsrc (easy)
2. Set the properties of the queue2 in the pipeline (easy)
3. Handle GST_MESSAGE_BUFFERING (easy to implement _a_ handler; unknown difficulty to actually "do the right thing" when low-threshold and high-threshold messages are received)
4. Figure out where to send the data when librespot pushes us a byte array (???)
5. Remove the static GstBufferPool usage since it'll be no longer needed, I assume (easy?)
6. Do we still need two Gst-specific threads (one for copying data into the buffer, and another for the actual mainloop) or can we / should we do buffer copying on the mainloop? (medium)

Reply:
@allquixotic I know really nothing about rust but I have experience with custom pipelines in python, so I know how to handle buffering in gstreamer. I'm in the process of creating a proper pipeline by looking at the rust binding examples that doesn't use parse latch. Even if I get it to work it will be ugly from a rust code perspective so it will need a clean up.

Reply:
> 
> 
> @allquixotic I know really nothing about rust but I have experience with custom pipelines in python, so I know how to handle buffering in gstreamer. I'm in the process of creating a proper pipeline by looking at the rust binding examples that doesn't use parse latch. Even if I get it to work it will be ugly from a rust code perspective so it will need a clean up.

OK, sounds great! Do you think we should just hack it by upping the buffer pool size to `4096 * format.size()` (which seems to be enough for all workloads that librespot currently throws at it) for the purpose of roderickd's PR, and do a separate PR to make gstreamer.rs less terrible from a gst design perspective?

Reply:
> OK, sounds great! Do you think we should just hack it by upping the buffer pool size to `4096 * format.size()` (which seems to be enough for all workloads that librespot currently throws at it) for the purpose of roderickd's PR, and do a separate PR to make gstreamer.rs less terrible from a gst design perspective?

If it works for now sure.

Do you actually use the Gstreamer backend? Does anyone? Or was it an example of someone thought it would be neat? I ask because I don't use it and don't want to be responsible for maintaining it and if no one actually uses it it should just be ejected from the code base.

Reply:
> 
> 
> > OK, sounds great! Do you think we should just hack it by upping the buffer pool size to `4096 * format.size()` (which seems to be enough for all workloads that librespot currently throws at it) for the purpose of roderickd's PR, and do a separate PR to make gstreamer.rs less terrible from a gst design perspective?
> 
> If it works for now sure.
> 
> Do you actually use the Gstreamer backend? Does anyone? Or was it an example of someone thought it would be neat? I ask because I don't use it and don't want to be responsible for maintaining it and if no one actually uses it it should just be ejected from the code base.

I use it every day, and I'm ... well, while I'm not the original author, I was responsible for getting it over the hump to get it merged to dev. So maybe it's just me :)

Reply:
> I use it every day

Why? What is your use case that you can't use the ALSA or another backend? Gstreamer in this case is just middle manning the PCM with no real benefit that I can see?

Reply:
If we want to avoid specifying the size of the gstbuffers in the bufferpool then I suggest we get rid of the bufferpool altogether. Yes, this always allocates a new buffer each time and will be technically less efficient.

```
--- a/playback/src/audio_backend/gstreamer.rs
+++ b/playback/src/audio_backend/gstreamer.rs
@@ -35,29 +35,20 @@ impl Open for GstreamerSink {
         let appsrc: gst_app::AppSrc = appsrce
             .dynamic_cast::<gst_app::AppSrc>()
             .expect("Couldn't cast AppSrc element at runtime!");
-        let bufferpool = gst::BufferPool::new();
-        let appsrc_caps = appsrc.get_caps().expect("Couldn't get appsrc caps");
-        let mut conf = bufferpool.get_config();
-        conf.set_params(Some(&appsrc_caps), 8192, 0, 0);
-        bufferpool
-            .set_config(conf)
-            .expect("Couldn't configure the buffer pool");
-        bufferpool
-            .set_active(true)
-            .expect("Couldn't activate buffer pool");
 
         let (tx, rx) = sync_channel::<Vec<u8>>(128);
         thread::spawn(move || {
             for data in rx {
-                let buffer = bufferpool.acquire_buffer(None);
+                let buffer = gst::Buffer::with_size(data.len());
                 if !buffer.is_err() {
                     let mut okbuffer = buffer.unwrap();
                     let mutbuf = okbuffer.make_mut();
-                    mutbuf.set_size(data.len());
```

Happy to defer this and anything else more to a separate more focused PR. The [proposed scaling fixes](https://github.com/librespot-org/librespot/pull/660#discussion_r596418821) are already fine for this PR.

Reply:
> 
> 
> > I use it every day
> 
> Why? What is your use case that you can't use the ALSA or another backend? Gstreamer in this case is just middle manning the PCM with no real benefit that I can see?

Nope, I use either the `pitch` processing element or the `shout2send` sink (and sometimes both) on a regular basis for tasks that no other librespot backend can do. I would very much prefer not to have to maintain this codebase myself, when there was at least a fair bit of interest in the issue that talked about gst support. 

I'm sure you might further question why I want to use either of these elements, but my answer is "because I like to" and hopefully that satisfies your curiosity. :)

This is getting _way_ into the weeds for this PR. But I guess I can't imagine that I'm the only Gstreamer backend user (or potential user) given the level of interest here that goes beyond just technical interest in doing it "because we can":

https://github.com/librespot-org/librespot/issues/28
https://github.com/plietar/librespot/issues/60
https://github.com/librespot-org/librespot/pull/456

Even if we do nothing about the design of the buffering for gst and just fix the sizing issues, the backend should still be restored to full functionality using any sample format for virtually all use cases of librespot going forward. I appreciate efforts to get it right for correctness and possible elimination of edge cases, but I don't think this audio backend is so bad or useless as to warrant culling from the tree. 

If anything, gst makes redundant certain other audio API backends that are basically thin wrappers around various "actual" (hardware) sound APIs, owing to the extremely wide-ranging support for gstreamer on anything from Raspberry Pi to Android to Windows/Mac/desktop Linux/BSD. Just `autoaudiosink` alone handles the problem of correctly supporting nearly all platform differences in audio backends for just about every device under the sun, while some libraries like portaudio don't receive nearly that level of wide ranging compatibility.

And unlike portaudio or some of those others, gst has a rich and varied set of high-quality processing plugins and sinks for "non-standard" things you might want to do with the audio coming out of librespot. Neither Gst itself, nor this audio backend, are bug-free; but especially in recent years you can scarcely find another media framework that is so widely supported on so many platforms that also supports a vast and diverse set of processing capabilities.

Reply:
> I use either the `pitch` processing element or the `shout2send` sink (and sometimes both) on a regular basis for tasks that no other librespot backend can do.

How do you use them with the backend as-is? `autoaudiosink` outputs to ALSA or PulseAudio or whatever on PC and mac right?

> "because I like to"

I can accept that as an answer. I mean really, that's the answer to why I use librespot,lol!!!

> I would very much prefer not to have to maintain this codebase myself, when there was at least a fair bit of interest in the issue that talked about gst support.

Well someone has to step up or the code will rot.

My vote (if it matters) is to do the bare minimum to get it working and then it would be advantageous to those that actually use the backend to do a more proper gstreamer implementation.

Reply:
Resolved in https://github.com/librespot-org/librespot/pull/660/commits/b94879de62f3fbf11a38aa2fa5df11b24e9998b8. Let's keep the GStreamer improvements flowing in a separate PR!

Reply:
> Well someone has to step up or the code will rot.
> 
> My vote (if it matters) is to do the bare minimum to get it working and then it would be advantageous to those that actually use the backend to do a more proper gstreamer implementation.

Agreed. I didn't mean to say that I'd prefer not to maintain the codebase _at all_ -- I will jump in and help -- but it's good to have help from others. :) It's one of those things where having the ability to share the workload with others is what I appreciate, not that I'm proposing that I contribute 0 work. Just wanted to clear that up. Carry on! :)